### PR TITLE
Private Service Endpoint

### PR DESCRIPTION
**Pull request by cdavid**

### CODE REVIEW

File: cmd/cluster/network/endpoint/create_endpoint.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+				logrus.Debugln("Security principals are set, attempting to create")
+			} else {
+				logrus.Fatalln("Security principals are not set and are mandatory for Private Service Endpoints.")
+			}
+			securityPrincipalsString, _ := cmd.Flags().GetString("security-principals")
+			securityPrincipalsList := strings.Split(securityPrincipalsString, ",")
+
+			allClusterRegions := clusterData.Info.ClusterRegionInfoDetails
+			desiredRegions := util.Filter(allClusterRegions, func(region ybmclient.ClusterRegionInfoDetails) bool {
+				return region.Region == reg
```
Comment (posriniv) [inline cmd/cluster/network/endpoint/create_endpoint.go:57]: nit: maybe make this regionInfo.Region == region ??

File: cmd/cluster/network/endpoint/create_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+			})
+
+			if len(desiredRegions) == 0 {
+				logrus.Fatalf("No region found for cluster %s with name %s", clusterData.Spec.Name, reg)
```
Comment (posriniv) [inline cmd/cluster/network/endpoint/create_endpoint.go:61]: nit: Whenever we are using Fatalf, maybe add a newline character at the end?

File: cmd/cluster/network/endpoint/delete_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+		default:
+			logrus.Fatalf("Endpoint is not a private service endpoint. Only private service endpoints are currently supported.")
+		}
+
+	},
+}
+
+func init() {
+	EndpointCmd.AddCommand(deleteEndpointCmd)
+	deleteEndpointCmd.Flags().String("endpoint-id", "", "[REQUIRED] THe ID of the endpoint")
```
Comment (posriniv) [inline cmd/cluster/network/endpoint/delete_endpoint.go:63]: can't we delete based on the region?

File: cmd/cluster/network/endpoint/update_endpoint.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		default:
+			logrus.Fatalf("Endpoint is not a private service endpoint. Only private service endpoints are currently supported.")
+		}
+
+	},
+}
+
+func init() {
+	EndpointCmd.AddCommand(updateEndpointCmd)
+	updateEndpointCmd.Flags().String("endpoint-id", "", "[REQUIRED] The ID of the endpoint")
```
Comment (posriniv) [inline cmd/cluster/network/endpoint/update_endpoint.go:83]: again, more of a question: if there is a single endpoint per region, can we use the region?

**Review by posriniv — APPROVED** [review]
Looks good to me!

File: cmd/cluster/network/endpoint/delete_endpoint.go (hunk 3 above)
Comment (cdavid) [reply cmd/cluster/network/endpoint/delete_endpoint.go:63]: The pair (Accessibility_type, region) should be unique. But for the first implementation, I think it's OK to require the endpoint id.

File: cmd/cluster/network/endpoint/update_endpoint.go (hunk 4 above)
Comment (cdavid) [reply cmd/cluster/network/endpoint/update_endpoint.go:83]: See above.

File: cmd/cluster/network/endpoint/create_endpoint.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		if err != nil {
+			logrus.Fatalf("Could not initiate api client: %s", err.Error())
+		}
+		authApi.GetInfo("", "")
+
+		clusterData := getCluster(cmd, authApi)
+		accessibilityType, _ := cmd.Flags().GetString("accessibility-type")
+		reg, _ := cmd.Flags().GetString("region")
+
+		switch accessibilityType {
```
Comment (shinji62) [inline cmd/cluster/network/endpoint/create_endpoint.go:45]: What is the point of the switch if only value ad default is possible ? `if` should be ok I guess

File: cmd/cluster/network/endpoint/util_endpoint.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	"encoding/json"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/yugabyte/ybm-cli/cmd/util"
+	ybmAuthClient "github.com/yugabyte/ybm-cli/internal/client"
+	ybmclient "github.com/yugabyte/yugabytedb-managed-go-client-internal"
+)
+
+func getCluster(cmd *cobra.Command, authApi *ybmAuthClient.AuthApiClient) ybmclient.ClusterData {
```
Comment (shinji62) [inline cmd/cluster/network/endpoint/util_endpoint.go:28]: This already exist in the authClient I think no ?

File: cmd/cluster/network/endpoint/util_endpoint.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	if len(resp.GetData()) == 0 {
+		logrus.Fatalf("Cluster not found\n")
+	}
+
+	return resp.GetData()[0]
+}
+
+func getEndpoints(cmd *cobra.Command, authApi *ybmAuthClient.AuthApiClient) ([]ybmclient.Endpoint, string) {
```
Comment (shinji62) [inline cmd/cluster/network/endpoint/util_endpoint.go:47]: Shoudll move to the client.go as well

File: cmd/cluster/network/endpoint/util_endpoint.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+	clusterId := clusterData.Info.Id
+	clusterEndpoints := clusterData.Info.ClusterEndpoints
+	jsonEndpoints, _ := json.Marshal(clusterEndpoints)
+	logrus.Debugf("Found endpoints: %v", string(jsonEndpoints))
+
+	return clusterEndpoints, clusterId
+}
+
+func getEndpointById(cmd *cobra.Command, authApi *ybmAuthClient.AuthApiClient) ([]ybmclient.Endpoint, string, string) {
```
Comment (shinji62) [inline cmd/cluster/network/endpoint/util_endpoint.go:58]: Same here should be in client.go as others

File: cmd/cluster/network/endpoint/util_endpoint.go (hunk 7 above)
Comment (shinji62) [reply cmd/cluster/network/endpoint/util_endpoint.go:47]: no need to pass the cmd in this case

File: cmd/cluster/network/endpoint/util_endpoint.go (hunk 8 above)
Comment (shinji62) [reply cmd/cluster/network/endpoint/util_endpoint.go:58]: no need to pass the all cmd just endpoint id should be enough

File: cmd/cluster/network/endpoint/create_endpoint.go (hunk 5 above)
Comment (cdavid) [reply cmd/cluster/network/endpoint/create_endpoint.go:45]: This is because we will add more cases in the future.

**Pull request merged**